service/tablet_allocator.cc

```diff
@@ -949,6 +962,26 @@ class load_balancer {
                         return std::make_pair(r1_it->second, r2[i]);
                     }
                 }
+                // After subtracting co-located replicas and exhausting intra-node migrations, we might
+                // decide to do rack-aware co-locations to minimize the amount of cross-rack migrations
+                // in order to not reduce availability at rack level.
+                // For example, replica sets of tablets r1 and r2 might be as follow (RF=racks=3):
+                // [r1: (rack1,rack1,rack2)], [r2: (rack1,rack3,rack1)]
+                // Without rack awareness, 2nd replica of r2 in rack3 could be migrated to co-habit 2nd
+                // replica of r1 in rack1, meaning r2 will lose availability since all its replicas will
+                // be on rack 1.
+                // With rack awareness, r1 and r2 replica sets are ordered by rack, while preserving the
+                // order by host id in each rack, to handle RF=N scenario. They'll look like this:
+                // [r1: (rack1,rack1,rack2)], [r2: (rack1,rack1,rack3)]
+                // In the best scenario, there's equal number of replicas in each rack, and co-location
+                // will not have to go across racks even a single time.
+                if (rack_aware_enabled) {
+                    auto cmp = [&topo = _tm->get_topology()] (const tablet_replica& r1, const tablet_replica& r2) {
+                        return topo.get_rack(r1.host) < topo.get_rack(r2.host);
+                    };
+                    std::ranges::stable_sort(r1, cmp);
+                    std::ranges::stable_sort(r2, cmp);
+                }
```

Comment:
Is it legal to just rearrange tablet replicas like this?

Reply:
Then let's sort permanently and keep it sorted.

If we need a notion of a primary owner,  we can add a new column primary_owner_in_datacenter map\<text, int>.

Reply:
off-topic: we should remember to only enable views on a table if we detect each tablet satisfy RF=#racks (not only at configuration level, but also if replica sets reflect the configuration).